consensus/src/dag/dag_fetcher.rs

```diff
@@ -0,0 +1,132 @@
+    oneshot,
+};
+
+/// Represents a request to fetch missing dependencies for `target`, `start_round` represents
+/// the first round we care about in the DAG, `exists_bitmask` is a two dimensional bitmask represents
+/// if a node exist at [start_round + index][validator_index].
+#[derive(Serialize, Deserialize, Clone)]
+struct FetchRequest {
+    target: NodeMetadata,
+    start_round: Round,
+    exists_bitmask: Vec<Vec<bool>>,
+}
+
+/// Represents a response to FetchRequest, `certified_nodes` are indexed by [round][validator_index]
+/// It should fill in gaps from the `exists_bitmask` according to the parents from the `target_digest` node.
+#[derive(Serialize, Deserialize, Clone)]
+struct FetchResponse {
+    epoch: u64,
+    certifies_nodes: Vec<Vec<CertifiedNode>>,
+}
+
+impl FetchResponse {
+    pub fn verify(
+        self,
+        _request: &FetchRequest,
+        _validator_verifier: &ValidatorVerifier,
+    ) -> anyhow::Result<Self> {
+        todo!("verification");
+    }
+}
+
+impl DAGMessage for FetchRequest {
+    fn epoch(&self) -> u64 {
+        self.target.epoch()
+    }
+}
+
+impl DAGMessage for FetchResponse {
+    fn epoch(&self) -> u64 {
+        self.epoch
+    }
+}
+
+enum FetchCallback {
+    Node(Node, oneshot::Sender<Node>),
+    CertifiedNode(CertifiedNode, oneshot::Sender<CertifiedNode>),
+}
+
+impl FetchCallback {
+    pub fn responders(&self, validators: &[Author]) -> Vec<Author> {
+        match self {
+            FetchCallback::Node(node, _) => vec![*node.author()],
+            FetchCallback::CertifiedNode(node, _) => node.certificate().signers(validators),
+        }
+    }
+
+    pub fn notify(self) {
+        if let Err(_) = match self {
+            FetchCallback::Node(node, sender) => sender.send(node).map_err(|_| ()),
+            FetchCallback::CertifiedNode(node, sender) => sender.send(node).map_err(|_| ()),
+        } {
+            error!("Failed to send node back");
+        }
+    }
+}
+
+struct DagFetcher {
+    epoch_state: Arc<EpochState>,
+    network: Arc<dyn DAGNetworkSender>,
+    dag: Arc<RwLock<Dag>>,
+    request_rx: Receiver<(FetchRequest, FetchCallback)>,
+}
+
+impl DagFetcher {
+    pub fn new(
+        epoch_state: Arc<EpochState>,
+        network: Arc<dyn DAGNetworkSender>,
+        dag: Arc<RwLock<Dag>>,
+    ) -> (Self, Sender<(FetchRequest, FetchCallback)>) {
+        let (request_tx, request_rx) = tokio::sync::mpsc::channel(16);
+        (
+            Self {
+                epoch_state,
+                network,
+                dag,
+                request_rx,
+            },
+            request_tx,
+        )
+    }
+
+    pub async fn start(mut self) {
+        while let Some((request, callback)) = self.request_rx.recv().await {
+            let responders =
+                callback.responders(&self.epoch_state.verifier.get_ordered_account_addresses());
+            let network_request = request.clone().into_network_message();
+            if let Ok(response) = self
```

Comment:
How do we tackle request deduplication? If we get two nodes in round (r) with missing nodes in r-1, those two node in round r will trigger a fetch, but I believe the fetcher should be smart enough to dedup the request otherwise in the worst case we might be sending fetch rpc for n nodes in worst case. 

also, we may need to check the dag store before sending the request, because the previous request might have already fetched the necessary nodes?

Do you want to add these as a TODO?

Reply:
what I have in mind is the fetcher is single task only, so it won't have all the concurrency issues you mentioned. Given the 2f+1 overlaps per round, the benefit of concurrent seems minimal to me. if there's node from much higher round, we'd fall into state sync instead.